tools/pip-compile-wrapper.py

```diff
@@ -0,0 +1,300 @@
+#!/usr/bin/env python
+"""
+Helper utility to compile / upgrade requirements files from templates.
+
+This only manages dependencies between requirements sources.
+The actual compiling is delegated to ``pip-compile`` from the ``pip-tools` package.
+
+NOTE: This utility must only use stdlib imports in order to be runnable even
+      before the dev requirements are installed.
+"""
+
+import os
+import re
+import shlex
+import subprocess
+import sys
+from collections import defaultdict
+from enum import Enum
+from itertools import chain, groupby, repeat
+from operator import itemgetter
+from optparse import OptionParser
+from pathlib import Path
+from shutil import which
+from typing import Dict, Iterable, Iterator, List, Optional, Set, Tuple
+
+REQUIREMENTS_SOURCE_DEV = "requirements-dev.in"
+REQUIREMENT_RE = re.compile(r"^([^=!~<>\[ ]+)")
+SCRIPT_NAME = os.environ.get("_SCRIPT_NAME", sys.argv[0])
+REQUIREMENTS_DIR = Path(__file__).parent.parent.joinpath("requirements").resolve()
+SOURCE_PATHS: Dict[str, Path] = {
+    path.relative_to(REQUIREMENTS_DIR).stem: path.resolve()
+    for path in REQUIREMENTS_DIR.glob("*.in")
+}
+TARGET_PATHS: Dict[str, Path] = {
+    name: REQUIREMENTS_DIR.joinpath(name).with_suffix(".txt") for name in SOURCE_PATHS.keys()
+}
+SOURCE_DEPENDENCIES: Dict[str, Set[str]] = defaultdict(set)
+
+
+class TargetType(Enum):
+    SOURCE = 1
+    TARGET = 2
+    ALL = 3
+
+
+def _resolve_source_dependencies() -> None:
+    for source_name, source_path in SOURCE_PATHS.items():
+        source_path = source_path.resolve()
+        # Dummy access to create the key
+        _ = SOURCE_DEPENDENCIES[source_name]
+        target_dir: Path = source_path.parent
+        with source_path.open("rt") as target_file:
+            for line in target_file:
+                line = line.strip()
+                if line.startswith("-r"):
+                    required = (
+                        target_dir.joinpath(line.lstrip("-r").strip())
+                        .resolve()
+                        .relative_to(REQUIREMENTS_DIR)
+                        .stem
+                    )
+                    SOURCE_DEPENDENCIES[source_name].add(required)
+
+
+_resolve_source_dependencies()
+
+
+def _run_pip_compile(
+    source_name: str,
+    upgrade_all: bool = False,
+    upgrade_packages: Optional[Set[str]] = None,
```

Comment:
are the options `upgrade_all` and `upgrade_packages` exclusive?

Edit: To answer my self, yes. This would probably be best done by two distinct functions.

Reply:
Hm, 95% of the code is identical between regular `compile`, `upgrade_packages` and `upgrade_all`. 
Not sure if refactoring that makes much sense.